File: codes/leetcode/KeysandRooms.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
class Solution {
private:
    void dfs(int index, vector<vector<int>>& rooms, vector<bool>& visited) {
        visited[index] = true; 
        for (int adjacent : rooms[index]) 
            if (!visited[adjacent])
                dfs(adjacent, rooms, visited); 
    }
    
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int numberOfRooms = rooms.size();
        vector<bool> visited(numberOfRooms, false);
        dfs(0, rooms, visited);
        for(bool status : visited) {
            if(!status)
                return false;
        }
        return true;
    }
};

File: codes/leetcode/KeysandRooms.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

from collections import deque

class Solution:

    def canVisitAllRooms(self, rooms: list[list[int]]) -> bool:
        n = len(rooms)
        visited = [False]*n
        queue = deque()

        i = 0
        queue.append(0)
        visited[0] = True
        while queue:
            i += 1
            node = queue.popleft()
            for neighbor in rooms[node]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)

        return (i==n)
